walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,235 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
+
+  /**
+   * Finds a list of potential meeting times based on the meeting request and 
+   * a list of existing events and their attendees. 
+   * Picks time slots if one or more time slots 
+   * exists so that both mandatory and optional attendees can attend. 
+   * Otherwise, returns the time slots that fit just the mandatory attendees.
+   */ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees(); 
+    long meetingDuration = request.getDuration(); 
+
+    List<TimeRange> mandatoryAttendeesBusyTimeRanges = 
+        getAllBusyTimeRanges(events, mandatoryAttendees); 
+    List<TimeRange> optionalAttendeesBusyTimeRanges = 
+        getAllBusyTimeRanges(events, optionalAttendees);
+
+    sortTimeRangesByStart(mandatoryAttendeesBusyTimeRanges); 
+    sortTimeRangesByStart(optionalAttendeesBusyTimeRanges);
+
+    List<TimeRange> mandatoryAttendeesFreeTimeRanges = 
+        getFreeTimeRanges(mandatoryAttendeesBusyTimeRanges, meetingDuration); 
+    List<TimeRange> optionalAttendeesFreeTimeRanges = 
+        getFreeTimeRanges(optionalAttendeesBusyTimeRanges, meetingDuration);
+
+    if (mandatoryAttendees.isEmpty()) {
+      return optionalAttendeesFreeTimeRanges; 
+    } else if (optionalAttendees.isEmpty()) {
+      return mandatoryAttendeesFreeTimeRanges; 
+    } else {
+      List<TimeRange> allFreeTimeRanges = findCommonFreeTimeRanges(
```

Comment:
(optional) Here you can just return findCommon... instead of declaring the variable.

Reply:
Updated.